app/jobs/update_from_saml_service.rb

```diff
@@ -0,0 +1,161 @@
+    org_name_node = xpath_at(node, './md:OrganizationDisplayName')
+
+    return nil unless org_id_node && org_name_node
+
+    org_id = org_id_node.content.strip
+    org_name = org_name_node.content.strip
+
+    org = Organization.find_or_initialize_by(domain: org_id)
+    org_attrs = { name: org_name }
+
+    org_attrs[:identifier] = org_identifier(org_id) if org.id.nil?
+
+    org.update!(org_attrs)
+
+    org
+  end
+
+  def process_idp(node, entity_id, org, reg_date)
+    idp = IdentityProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
+    name = name_node.content.strip if name_node
+
+    idp.update!(name: name, organization: org)
+
+    idp_attributes = idp.identity_provider_saml_attributes
+    process_idp_attributes(idp_attributes, node)
+
+    @identity_providers.append(idp)
+    activate_object(idp, reg_date)
+  end
+
+  def process_sp(node, entity_id, org, reg_date)
+    sp = ServiceProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
+    name = name_node.content.strip if name_node
+
+    sp.update!(name: name, organization: org)
+
+    sp_attributes = sp.service_provider_saml_attributes
+    process_sp_attributes(sp_attributes, node)
+
+    @service_providers.append(sp)
+    activate_object(sp, reg_date)
+  end
+
+  def process_idp_attributes(scope, node)
+    nodes = xpath(node, './saml:Attribute')
+    process_attributes(scope, nodes) do |_, assoc|
+      assoc.save
+    end
+  end
+
+  def process_sp_attributes(scope, node)
+    nodes = xpath(node, './md:AttributeConsumingService/md:RequestedAttribute')
+    process_attributes(scope, nodes) do |attr, assoc|
+      is_req = attr_val(attr, 'isRequired') == 'true'
+      assoc.update(optional: !is_req)
+    end
+  end
+
+  def process_attributes(scope, nodes)
+    attrs = nodes.map do |attr|
+      attr_name = attr_val(attr, 'FriendlyName')
+      attribute = SAMLAttribute.find_by(name: attr_name)
+      next unless attribute
+
+      assoc = scope.find_or_initialize_by(saml_attribute_id: attribute.id)
+
+      assoc if yield attr, assoc
+    end
+
+    # Delete attributes that aren't associated with this IdP/SP any more
+    scope.where.not(id: attrs.compact.map(&:id)).destroy_all
+  end
+
+  def activate_object(obj, date)
+    obj.activations.find_or_initialize_by({}).update!(activated_at: date)
+  end
+
+  def source
+    configuration[:metadata_url]
+  end
+
+  def metadata_cert
+    cert = configuration[:metadata_cert]
+    OpenSSL::X509::Certificate.new(cert) if cert
+  end
+
+  def configuration
+    Rails.application.config.reporting_service.saml_service
+  end
+
+  def org_identifier(name)
+    # Generate a valid identifier for an organisation given the OrganisationName (domain)
+    digest = OpenSSL::Digest.new('SHA256').digest("tuakiri:subscriber:#{name}")
```

Comment:
I'm a little concerned about having two `org_identifier` methods that provide the same function but implement it differently - potential source for confusion.
https://github.com/REANNZ/reporting-service/blob/master/app/jobs/concerns/query_federation_registry.rb#L13

Reply:
Renamed to avoid confusion.